rust/main/agents/relayer/src/server/message_retry.rs

```diff
@@ -1,32 +1,89 @@
+use std::sync::Arc;
+
 use crate::settings::matching_list::MatchingList;
+
 use axum::{extract::State, routing, Json, Router};
+
 use derive_new::new;
-use tokio::sync::broadcast::Sender;
+use serde::{Deserialize, Serialize};
+use tokio::sync::{broadcast::Sender, mpsc, Mutex};
 
 const MESSAGE_RETRY_API_BASE: &str = "/message_retry";
 
-#[derive(new, Clone)]
+#[derive(new)]
 pub struct MessageRetryApi {
-    tx: Sender<MatchingList>,
+    tx: Sender<MessageRetryRequest>,
+    rx: Arc<Mutex<mpsc::Receiver<MessageRetryResponse>>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryApiState {
+    pub tx: Sender<MessageRetryRequest>,
+    pub rx: Arc<Mutex<mpsc::Receiver<MessageRetryResponse>>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryRequest {
+    pub uuid: String,
+    pub pattern: MatchingList,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MessageRetryResponse {
+    /// ID of the retry request
+    pub uuid: String,
+    /// how many pending operations were processed
+    pub processed: usize,
+    /// how many of the pending operations matched the retry request pattern
+    pub matched: u64,
 }
 
 async fn retry_message(
-    State(tx): State<Sender<MatchingList>>,
+    State(state): State<MessageRetryApiState>,
     Json(retry_req_payload): Json<MatchingList>,
-) -> String {
-    match tx.send(retry_req_payload) {
-        Ok(_) => "Moved message(s) to the front of the queue".to_string(),
-        // Technically it's bad practice to print the error message to the user, but
-        // this endpoint is for debugging purposes only.
-        Err(err) => format!("Failed to send retry request to the queue: {}", err),
+) -> Result<Json<MessageRetryResponse>, String> {
+    let uuid = uuid::Uuid::new_v4();
+    let uuid_string = uuid.to_string();
+
+    tracing::debug!("Sending message retry request: {uuid_string}");
+
+    state
+        .tx
+        .send(MessageRetryRequest {
+            uuid: uuid_string.clone(),
+            pattern: retry_req_payload,
+        })
+        .map_err(|err| {
+            // Technically it's bad practice to print the error message to the user, but
+            // this endpoint is for debugging purposes only.
+            format!("Failed to send retry request to the queue: {}", err)
+        })?;
+
+    let mut rx = state.rx.lock().await;
+
+    // Wait for response from message retry.
+    // Warning: this potentially blocks other retry requests
+    // from properly returning because their response might be consumed by
+    // another request
+    loop {
```

Comment:
especially as we have a lock acquired, can we have a timeout? It can be long, like 60s or more, but want to avoid waiting forever if something goes wrong